### PR TITLE
PP-684 Enable self service card type restrictions from Staff Pages

### PR DESCRIPTION
**Pull request by ckalista**
## WHAT
- Changesets in migration script for new tables card_types and
  accepted_card_types.
- CardTypeDao implementing findById, findAll.
- GatewayAccountEntity updated to embed a collection of card types with
  JPA mapping accordingly to card_types and accepted_card_types tables.
- CardTypesResource exposing all card types available in the system.
- CardTypesResource exposing all card types accepted for a given
  account and allowing to update card types accepted.
- UuidAbstractEntity used as a base class for CardTypeEntity to map the
  primary key to a UUID instead of a long.
- Unit tests and integration tests.
## WHO
- [x] Developers
- [] WebOps
## Related PRs

_Related PRs against other branches:_

| branch | PR |
| --- | --- |
| other_branch | [link]() |

### CODE REVIEW

File: src/main/java/uk/gov/pay/connector/model/domain/CardTypeEntity.java [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+
+        public static Type fromString(String type) {
+            for (Type t : values()) {
+                if (StringUtils.equals(t.getValue(), type)) {
+                    return t;
+                }
+            }
+            throw new IllegalArgumentException("type not recognized: " + type);
+        }
+    }
```
Comment (rahul-vinayak) [inline src/main/java/uk/gov/pay/connector/model/domain/CardTypeEntity.java]: You could probably make it simpler if we don't do 

> CREDIT("CREDIT")

and just go 

> CREDIT

File: src/main/java/uk/gov/pay/connector/model/domain/UUIDConverter.java [hunk 2]
```diff
@@ -0,0 +1,17 @@
+package uk.gov.pay.connector.model.domain;
+
+import javax.persistence.AttributeConverter;
+import java.util.UUID;
+
+@javax.persistence.Converter(autoApply = true)
+public class UuidConverter implements AttributeConverter<UUID, UUID> {
+    @Override
```
Comment (rahul-vinayak) [inline src/main/java/uk/gov/pay/connector/model/domain/UUIDConverter.java]: the class name and the file name are a mismatch (Case), and Intellij complains about it

File: src/main/java/uk/gov/pay/connector/resources/CardTypesResource.java [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+    public CardTypesResource(CardTypeDao cardTypeDao) {
+        this.cardTypeDao = cardTypeDao;
+    }
+
+    @GET
+    @Path(CARD_TYPES_API_PATH)
+    @Produces(APPLICATION_JSON)
+    @Transactional
+    public Response getCardTypes() {
+        return successResponseWithEntity(ImmutableMap.of(CARD_TYPES_FIELD_NAME, cardTypeDao.findAll()));
```
Comment (rahul-vinayak) [inline src/main/java/uk/gov/pay/connector/resources/CardTypesResource.java]: convention we are following is json goes out in snake_case, so should be card_types

File: src/main/java/uk/gov/pay/connector/resources/GatewayAccountResource.java [hunk 4]
```diff
@@ -64,6 +70,33 @@ public Response getGatewayAccount(@PathParam("accountId") Long accountId) {
...
+    @GET
+    @Path(FRONTEND_ACCOUNT_CARDTYPES_API_PATH)
+    @Consumes(APPLICATION_JSON)
+    @Produces(APPLICATION_JSON)
+    public Response getGatewayAccountAcceptedCardTypes(@PathParam("accountId") Long accountId) {
+        logger.info("Getting accepted card types for gateway account with account id {}", accountId);
+        return gatewayDao
+                .findById(GatewayAccountEntity.class, accountId)
+                .map(gatewayAccount -> successResponseWithEntity(ImmutableMap.of(CARD_TYPES_FIELD_NAME, gatewayAccount.getCardTypes())))
+                .orElseGet(() -> notFoundResponse(format("Account with id %s not found.", accountId)));
```
Comment (rahul-vinayak) [inline src/main/java/uk/gov/pay/connector/resources/GatewayAccountResource.java]: again json keys goes out as snake_case

File: src/main/java/uk/gov/pay/connector/resources/GatewayAccountResource.java [hunk 5]
```diff
@@ -158,6 +177,39 @@ public Response updateGatewayAccountServiceName(@PathParam("accountId") Long gat
...
+
+        List<UUID> cardTypeIds = cardTypes.get(CARD_TYPES_FIELD_NAME);
+
+        List<UUID> cardTypeIdNotFound = cardTypeIds.stream()
+                .filter(cardTypeId -> !cardTypeDao.findById(cardTypeId).isPresent())
+                .collect(Collectors.toList());
+
+        if (cardTypeIdNotFound.size() > 0) {
+            return badRequestResponse(format("CardType referenced by id '%s' does not exist", cardTypeIdNotFound.get(0)));
+        }
```
Comment (rahul-vinayak) [inline src/main/java/uk/gov/pay/connector/resources/GatewayAccountResource.java]: returning only 0th element?

> cardTypeIdNotFound.get(0)

File: src/main/java/uk/gov/pay/connector/model/domain/UUIDConverter.java (hunk 2 above)
Comment (ckalista) [inline src/main/java/uk/gov/pay/connector/model/domain/UUIDConverter.java]: Fixed!

File: src/main/java/uk/gov/pay/connector/resources/CardTypesResource.java (hunk 3 above)
Comment (ckalista) [inline src/main/java/uk/gov/pay/connector/resources/CardTypesResource.java]: Changed it to comply to our standard.

File: src/main/java/uk/gov/pay/connector/resources/GatewayAccountResource.java (hunk 4 above)
Comment (ckalista) [inline src/main/java/uk/gov/pay/connector/resources/GatewayAccountResource.java]: Changed it to comply.

File: src/main/java/uk/gov/pay/connector/resources/GatewayAccountResource.java (hunk 5 above)
Comment (ckalista) [inline src/main/java/uk/gov/pay/connector/resources/GatewayAccountResource.java]: Yes, we follow the fail fast principle we introduced recently in all our APIs, hence we return only the first field that doesn't pass validation.

Comment (rahul-vinayak) [inline src/main/java/uk/gov/pay/connector/resources/GatewayAccountResource.java]: Yes I agree fail-fast is a good approach when we don't want to do certain checks if something has failed, however if we already have a list of cardTypeIdNotFound available, so I would return the entire list and save the round trips?

Comment (ckalista) [inline src/main/java/uk/gov/pay/connector/resources/GatewayAccountResource.java]: Code has been updated as suggested to reference all the faulty ids in the error message.

File: src/main/java/uk/gov/pay/connector/model/domain/CardTypeEntity.java (hunk 1 above)
Comment (ckalista) [inline src/main/java/uk/gov/pay/connector/model/domain/CardTypeEntity.java]: Updated as suggested.

**Comment by ckalista** [comment]
rebuild

**Comment by ckalista** [comment]
rebuild

**Comment by ckalista** [comment]
rebuild

**Comment by ckalista** [comment]
rebuild

**Comment by ckalista** [comment]
rebuild

**Comment by PauloPortugal** [comment]
rebuild

**Pull request merged**